[Merged by Bors] - feat(tactic/norm_swap): simplify numeral swaps

pechersky opened the pull request:
Explicitly applied swap equalities over `nat` can be proven to be true using `dec_trivial`. However, if occurring in the middle of a larger expression, a separate specialized hypothesis would be necessary to reduce them down. This is an initial attempt at a `norm_num` extension which seeks to reduce down expressions of the `swap X Y Z` form. Handles `nat`, `int`, `rat`, and `fin` swaps.

---
<!--
put comments you want to keep out of the PR commit here.
If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

- [x] depends on: #5820 

For posterity, the PR's initial comment:

> The lemmas added here help to simplify any applied swap where the swappands are even (`bit0`) and the applied value is odd (`bit1`) or vice versa. Special instantiations of the lemmas are needed for the `0` and `1` base cases. With these, some swaps of `nat` numerals will be simplified. This also provides a path forward for simplification of swaps over `fin n`, which come up in sums over permutations of finite types.
> 
> Should there be more lemmas simplifying the numeral based swaps? For example, ones of the form
> ```
> swap (bit0 x) (bit1 y) (bit0 z) = ite (x = z) (bit1 y) (bit0 z)
> ```
> Even for those, there would still need to be some other simplifying lemmas about how to simplify propositions of `x = z`. That means lemmas about the injectivity of `bit0` and `bit1`, and simplifying `iff`s.
> 
> In total, there would be 4^3 (0, 1, bit0, bit1) lemmas about swap necessary. One could also write a metaprogram to insert these lemmas via a declaration.

pechersky committed:
feat(data/equiv/basic): simplify nat numeral swaps

pechersky committed:
add test file to make sure simp works on nat swaps

pechersky committed:
introduce `norm_swap` extension

This is an initial attempt at a `norm_num` extension which seeks
to reduce down expressions of the `swap X Y Z` form.
Right now, it only handles `nat`s.

pechersky committed:
delete the bit[01] swap lemmas defined

pechersky committed:
switch `trace` of `nat` limitation to `fail`

pechersky committed:
add copyright header and docstring

pechersky committed:
remove `name_of_const_or_app` and clean up

cipher1024 commented on src/tactic/norm_swap.lean:
```diff
@@ -0,0 +1,80 @@
+    let fname : name := fexpr.get_app_fn.const_name,
+    if fname = f then pure (α, inst, fexpr, x)
+    else
+      fail $ "retrieved function name " ++ fname.to_string ++ " is not the expected " ++ f.to_string
```
consider using `fail!"retrieved function name {fname} is not the expected {f}"
```suggestion
      fail!"retrieved function name {fname} is not the expected {f}"
```

pechersky committed:
use fail! string interpolation

Co-authored-by: Simon Hudon <simon.hudon@gmail.com>

pechersky committed:
slight spacing change

cipher1024 commented on src/tactic/norm_swap.lean:
```diff
@@ -0,0 +1,79 @@
+open equiv tactic expr
+
+/--
+If `e : expr` is of the form `app (⇑f) x`, get back the `expr`s for type of `f`,
```
I think this formulation might be more straightforward: 

```lean
/--
Match a term of shape `app (⇑f) x` and return:
  * the type `tf` of `f`
  * the instance `has_coe_to_fun tf`
  * `f`
  * `x`
  -/

cipher1024 replied:
In general, I see that you use phrasing like "the `expr` which is the *something something*". I think you could stick to saying "the *something something*" adding parenthetical notes if there's any possible ambiguity.

pechersky committed:
clarify docstrings

robertylewis commented on src/tactic/norm_swap.lean:
```diff
@@ -0,0 +1,78 @@
+/-!
+# `norm_swap`
+
+Evaluating `swap x y z` for numerals `x y z : ℕ`.
```
Could you write a more detailed doc here? Explain that this is a `norm_cast` plugin and how the evaluation works? I know it's simple, but two lines of English are faster to read than 10 lines of code to get the gist.

robertylewis commented on src/tactic/norm_swap.lean:
```diff
@@ -0,0 +1,78 @@
+  * `f`
+  * `x`
+  -/
+meta def expr.get_of_coe_fn (e : expr) (f : name) : tactic (expr × expr × expr × expr) :=
```
This doesn't need to be `tactic` valued and isn't specific to this file. It can live in `meta/expr.lean`  and be named `expr.is_app_coe_fun`.
```lean
meta def expr.get_of_coe_fn : expr → name → option (expr × expr × expr × expr)
| (app `(@coe_fn %%α %%inst %%fexpr) x) f :=
    let fname : name := fexpr.get_app_fn.const_name in
    if fname = f then (α, inst, fexpr, x) else none
| _ _ := none
```

eric-wieser replied:
If its not `tactic`-valued then there's no way to determine a suitable error message, is there?

robertylewis replied:
No, but the error message is eaten by `norm_num` anyway in this application. Many of these matching functions have `option`-valued `is_` and `tactic`-valued `match_` variants. But the match error message is often not what you want your tactic error message to look like.

robertylewis commented on src/tactic/norm_swap.lean:
```diff
@@ -0,0 +1,78 @@
+Creates the application `n c.α p l`, where `p` is a type class instance found in the cache `c`,
+but works on `c.α : Sort*` instead of how `mk_app` uses `c.α : Type*`.
+-/
+meta def tactic.instance_cache.mk_sorted_app (c : instance_cache) (n : name) (l : list expr) :
```
This also isn't specific to this file at all, it can move somewhere with other `instance_cache` stuff. Without looking up the implementation I'm not sure -- is this really a different function, or a generalization? Can it replace `mk_app`?

pechersky replied:
The problem is in `mk_instance_cache`:
```
/-- Creates an `instance_cache` for the type `α`. -/
meta def mk_instance_cache (α : expr) : tactic instance_cache :=
do u ← mk_meta_univ,
   infer_type α >>= unify (expr.sort (level.succ u)),
   u ← get_univ_assignment u,
   return ⟨α, u, mk_name_map⟩
```
There, the stored `instance_cache.univ` is 1 less than what is needed in my case. That is, `mk_instance_cache` assumes that any use of the cache via `mk_app` will be with functions that expect a `Type*`. Since `coe_fn` is over `Sort*`, `mk_app` does not work for me.

@robertylewis would you suggest editing how the `instance_cache` works, or just avoiding it all together and making the raw `app` `expr`?

robertylewis replied:
I think the proper solution (which belongs in another PR) is to change `instance_cache.mk_app` to check whether the constant expects a `Type` or `Sort`. e.g.

```lean
open tactic 

meta def takes_type_arg (n : name) : tactic bool := 
do d ← get_decl n, 
match d.type with 
| expr.pi _ _ (expr.sort (level.succ _)) _ := return tt
| expr.pi _ _ (expr.sort _) _ := return ff
| _ := failed
end 

universe u
def f {α : Type u} := ℕ
def g {α : Sort u} := ℕ

#eval takes_type_arg `f >>= trace
#eval takes_type_arg `g >>= trace
```

robertylewis replied:
That said: you don't need to use a `instance_cache` for `coe_fn` at all in your code. The cache will never be populated with a `has_coe_to_fun` before you use it (`prove_ne` won't add this instance) and it's discarded right after. Just use regular `tactic.mk_app`.

robertylewis commented on src/tactic/norm_swap.lean:
```diff
@@ -0,0 +1,78 @@
+
+/--
+Creates the application `n c.α p l`, where `p` is a type class instance found in the cache `c`,
+but works on `c.α : Sort*` instead of how `mk_app` uses `c.α : Type*`.
```
```suggestion
but works on `c.α : Sort*` instead of how `tactic.instance_cache.mk_app` uses `c.α : Type*`.
```
for linkification

robertylewis commented on src/tactic/norm_swap.lean:
```diff
@@ -0,0 +1,78 @@
+namespace norm_swap
+
+/--
+A tactic for normalizing `equiv.swap a b c` where `a b c : ℕ` are numerals.
```
```suggestion
A `norm_num` plugin for normalizing `equiv.swap a b c` where `a b c : ℕ` are numerals.
```

robertylewis commented on src/tactic/norm_swap.lean:
```diff
@@ -0,0 +1,78 @@
+  (swapt, coe_fn_inst, f, c) ← expr.get_of_coe_fn e ``equiv.swap,
+  [α, deceq_inst, a, b] ← pure f.get_app_args, -- the swap should have exactly two arguments applied
+  unify α `(ℕ) <|> (fail "currently, norm_swap supports only ℕ"),
+  dic ← mk_instance_cache α,
```
The instance cache is used at most once below, right? Or am I misreading? Using it doesn't save you anything over `mk_app` (and adds a small amount of overhead).

robertylewis commented on src/tactic/norm_swap.lean:
```diff
@@ -0,0 +1,78 @@
+  dic ← mk_instance_cache α,
+  na ← a.to_nat,
+  nb ← b.to_nat,
+  nc ← c.to_nat,
```
I think the only reason you're restricting yourself to `nat` is because you use `to_nat` here instead of `to_rat`. This should work on any type the `prove_ne` works on.

eric-wieser replied:
How hard would it be to make this work on `fin n` too?

pechersky replied:
That's exactly my goal, to make it work for `fin`. I don't think `to_rat` is good enough for it yet, because of the modular arithmetic. I want to make a `norm_fin` extension first that will normalize things that look like `2.succ`. Once that works, then we'll be able to combine them.

robertylewis replied:
Indeed, `to_rat` won't get you to `fin` since `norm_num` doesn't prove `fin` disequalities (yet). It does get you to the `int` example, with a couple other superficial changes.

pechersky committed:
clarify main docstring

pechersky committed:
generalize out `match_app_coe_fn`

pechersky committed:
no need to use instance_cache to generate proofs

pechersky committed:
docstring clarification

pechersky committed:
make `prove_lt_rat` work when `a` and `b` both negative

previously, it flipped the signs for both but didn't reverse the order
of the two when passing to the next proof constructor

pechersky committed:
remove extra mk_apps and work with int/rats

pechersky committed:
include nonneg int example and two small neg int and rat examples

pechersky committed:
broken neg int example

pechersky committed:
Revert "broken neg int example"

This reverts commit 7dbec44d572261bb5c76bdae7d4e7d01e716ec7d.

pechersky committed:
satisfy linter

pechersky committed:
broken neg int example

pechersky committed:
fix int example, provide rat also

eric-wieser commented on src/tactic/norm_swap.lean:
```diff
@@ -0,0 +1,56 @@
+  if ha : nc = na
+    then do
+      let p : expr := `(@swap_apply_left.{1} %%α %%deceq_inst %%a %%b),
+      pure (b, p)
```
Does this golf to
```suggestion
    then
      let p : expr := `(@swap_apply_left.{1} %%α %%deceq_inst %%a %%b) in
      (b, p)
```
I have no idea which is better style though

eric-wieser commented on src/tactic/norm_swap.lean:
```diff
@@ -0,0 +1,56 @@
+    then do
+      let p : expr := `(@swap_apply_left.{1} %%α %%deceq_inst %%a %%b),
+      pure (b, p)
+  else if hb : nc = nb
```
Am I right in thinking you don't use `hb` or `ha` at all below? If so,
```suggestion
  else if nc = nb
```

pechersky committed:
clean up statements

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

pechersky committed:
expand docstring to include int and rat

pechersky committed:
fix typo in docstring

pechersky committed:
first go at normalizing fin values

pechersky committed:
satisfy linter

pechersky committed:
feat(tactic/norm_fin): normalize fin numerals

pechersky committed:
line linter

pechersky committed:
simplify cast_succ proof

pechersky committed:
expand to_nat to deal with `+ 1`

pechersky committed:
make iterated norm_fin work, and on `n + 1` types

pechersky committed:
confirm via tests

pechersky committed:
add examples of in situ normalization

pechersky committed:
clarify comment

pechersky committed:
move lemmas to above section

pechersky committed:
provide fin.prove_lt

pechersky committed:
launder fin n through `eval_field`

pechersky committed:
remove out of place docstring

pechersky committed:
fix up docstrings

pechersky committed:
move lemmas into norm_num namepsace

pechersky committed:
remove instance_cache from prove_lt_fin

pechersky committed:
simplify numeral check in to_fin_one

pechersky committed:
rename helper lemmas

pechersky committed:
use eval_expr instead of to_nat

pechersky committed:
revert to_nat change

pechersky committed:
edit docstring with ticks

pechersky committed:
use eval_expr in prove_lt_fin

pechersky committed:
add test for not fully evaluated n in fin n type

digama0 committed:
feat(tactic/norm_fun): tactic for normalizing `fin n` expressions

This is based on #5791, with a new implementation using the
`normalize_fin` function.

digama0 committed:
docs

pechersky committed:
le lt ge gt tests

pechersky committed:
ne tests, that fail

pechersky committed:
initial ne fix

pechersky committed:
ne test with intro, fails

digama0 committed:
fix inequality handling

digama0 committed:
oops

pechersky committed:
remove cruft

pechersky committed:
add back tests

pechersky committed:
Merge remote-tracking branch 'origin/mario/norm-fin' into HEAD

pechersky committed:
clarify tests

pechersky committed:
implement prove_ne_fin based on mario's PR

pechersky committed:
Merge remote-tracking branch 'origin/mario/norm-fin' into pechersky/equiv-swap-nat-bit

pechersky committed:
use `prove_eq_ne_fin` instead

digama0 committed:
extract `prove_eq_ne_fin`

robertylewis committed:
Merge branch 'master' into pechersky/equiv-swap-nat-bit

robertylewis commented on test/norm_swap.lean:
```diff
@@ -0,0 +1,95 @@
+open equiv tactic
+
+local attribute [-simp] swap_apply_left
+local attribute [-simp] swap_apply_right
```
Are these just needed because of how the test is set up? (If so could you add a short comment?) Or are they needed to make the plugin work in practice?

pechersky replied:
Turning off the simp tags makes sure we're testing the tactic, and not the simp-using capabilities of `norm_num`. Comment added.

robertylewis commented on src/tactic/norm_swap.lean:
```diff
@@ -0,0 +1,54 @@
+  nb ← b.to_rat <|> (do (fb, _) ← norm_fin.eval_fin_num b, fb.to_rat),
+  nc ← c.to_rat <|> (do (fc, _) ← norm_fin.eval_fin_num c, fc.to_rat),
+  if nc = na
+    then let p : expr := `(@swap_apply_left.{1} %%α %%deceq_inst %%a %%b) in pure (b, p)
```
It looks like you're hardcoding `α : Type` (= `Sort 1`) here and below. Is there a reason not do do ``p ← mk_mapp `swap_apply_left [α, deceq_inst, a, b]`` and let Lean infer the universe? I guess this is most often used for `nat`/`rat`/etc in `Type`, but afaict nothing is stopping it from working on `α : Type u` with enough algebraic structure, right?

pechersky replied:
Implemented. In general, is `mk_mapp` faster because we provide all the terms that would have to be inferred if we had used `mk_app`?

robertylewis replied:
Yes, in particular it won't fire type class inference. Since you already have all the arguments available here we might as well make use of them.

pechersky committed:
clarify why untagging simp lemmas

pechersky committed:
generalize to any universe via mk_mapp
